packages/ws/src/strategies/sharding/WorkerShardingStrategy.ts

```diff
@@ -0,0 +1,211 @@
+
+	private readonly throttler: IdentifyThrottler;
+
+	public constructor(manager: WebSocketManager, options: WorkerShardingStrategyOptions) {
+		this.manager = manager;
+		this.throttler = new IdentifyThrottler(manager);
+		this.options = options;
+	}
+
+	public async spawn(shardIds: number[]) {
+		const shardsPerWorker = this.options.shardsPerWorker === 'all' ? shardIds.length : this.options.shardsPerWorker;
+
+		let shards = 0;
+		while (shards !== shardIds.length) {
+			const slice = shardIds.slice(shards, shardsPerWorker + shards);
+			const workerData: WorkerData = {
+				...(await managerToFetchingStrategyOptions(this.manager)),
+				shardIds: slice,
+			};
+
+			const worker = new Worker(join(__dirname, 'worker.js'), { workerData });
+			await once(worker, 'online');
+			worker
+				.on('error', (err) => {
+					throw err;
+				})
+				.on('messageerror', (err) => {
+					throw err;
+				})
+				// eslint-disable-next-line @typescript-eslint/no-misused-promises
+				.on('message', (payload: WorkerRecievePayload) => this.onMessage(worker, payload));
+
+			this.workers.push(worker);
+			for (const shardId of slice) {
+				this.workerByShardId.set(shardId, worker);
+			}
+
+			shards += slice.length;
+		}
+	}
+
+	public async connect() {
+		const promises = [];
+
+		for (const [shardId, worker] of this.workerByShardId.entries()) {
+			await this.throttler.waitForIdentify();
+
+			const payload: WorkerSendPayload = {
+				op: WorkerSendPayloadOp.Connect,
+				shardId,
+			};
+
+			const promise = new Promise<void>((resolve) => this.connectPromises.set(shardId, resolve));
+			worker.postMessage(payload);
+			promises.push(promise);
+		}
+
+		await Promise.all(promises);
+	}
+
+	public async destroy(options: Omit<WebSocketShardDestroyOptions, 'recover'> = {}) {
+		const promises = [];
+
+		for (const [shardId, worker] of this.workerByShardId.entries()) {
+			const payload: WorkerSendPayload = {
+				op: WorkerSendPayloadOp.Destroy,
+				shardId,
+				options,
+			};
+
+			const promise = new Promise<void>((resolve) => this.destroyPromises.set(shardId, resolve));
+			worker.postMessage(payload);
+			promises.push(promise);
+
+			await worker.terminate();
+		}
+
+		this.workers = [];
+		this.workerByShardId.clear();
+
+		await Promise.all(promises);
+	}
+
+	public send(shardId: number, data: GatewaySendPayload) {
+		const worker = this.workerByShardId.get(shardId);
+		if (!worker) {
+			throw new Error(`No worker found for shard ${shardId}`);
+		}
+
+		const payload: WorkerSendPayload = {
+			op: WorkerSendPayloadOp.Send,
+			shardId,
+			payload: data,
+		};
+		worker.postMessage(payload);
+	}
+
+	private async onMessage(worker: Worker, payload: WorkerRecievePayload) {
+		switch (payload.op) {
+			case WorkerRecievePayloadOp.Connected: {
+				const resolve = this.connectPromises.get(payload.shardId);
+				if (!resolve) {
+					console.warn(`No connect promise found for shard ${payload.shardId}`);
+				}
+
+				resolve!();
```

Comment:
Doing like this again will throw because `resolve` is `undefined`.
```suggestion
				if (!resolve) {
					console.warn(`No connect promise found for shard ${payload.shardId}`);
				}

				resolve?.();
```

Reply:
Talked about this Vlad, it'll be staying as-is.